feat(cast) add creation-code method [#8973]

pawurb opened the pull request:
## Motivation

As described in https://github.com/foundry-rs/foundry/issues/8973, currently there's no simple way to obtain contract creation code without compiling them locally. It is needed to deploy contracts with `sol!` macro. 

If this is accepted I'll add `artifact` method generating file that can be used directly with `sol!` macro. 

## Solution

This PR adds `cast creation-code` method. It outputs the contract creation code.

## Example usage

```
cast creation-code 0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2 --etherscan-api-key API_KEY --rpc-url RPC_URL
```

mattsse commented on crates/cast/bin/cmd/creation_code.rs:
```diff
@@ -0,0 +1,96 @@
+async fn fetch_creation_code(
+    contract: Address,
+    client: Client,
+    provider: &Arc<RetryProvider>,
```
```suggestion
    provider: &RetryProvider,
```

mattsse commented on crates/cast/bin/cmd/creation_code.rs:
```diff
@@ -0,0 +1,96 @@
+        // Extract creation code from tx traces
+        let mut creation_bytecode = None;
+
+        let traces = provider.trace_transaction(creation_tx_hash).await?;
```
this will work, but is usually not available on the free tier, so this can fail with a serde rpc error "method not found" or "unavailable", we should add an eyre context here like "failed to trace tx 0x..."

mds1 commented on crates/cast/bin/cmd/creation_code.rs:
```diff
@@ -0,0 +1,96 @@
+    }
+}
+
+/// Fetches the creation code of a contract from Etherscan and RPC.
```
We should document whether or not this includes constructor args appended to initcode

pawurb replied:
@mds1 right, I should probably strip them to make the bytecode useful for local deployment. Maybe worth returning constructor args separately?

pawurb replied:
I'm reading up there's is no a single unified convention for encoding constructor args with different solidity versions. So I think I'll just add a comment that they are appended.

mds1 replied:
What is the use case for this method? I think that impacts whether or not you want to strip the constructor args. If you do want to strip them you'll likely need to use blockscout or etherscan to fetch constructor args

pawurb commented:
I've added `--only-args` and `--without-args` flags to `creation-code` and `creation-args` method. With these tools it's possible to get bytecode for local deployment and also inspect args that were used in the original contract deployment.

```
cast creation-code 0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419 --only-args --etherscan-api-key API_KEY --rpc-url RPC_URL

# 0x000000000000000000000000f79d6afbb6da890132f9d7c355e3015f15f3406f0000000000000000000000000000000000000000000000000000000000000000
```

```
cast creation-args 0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419  --etherscan-api-key API_KEY --rpc-url RPC_URL

# address 0x000000000000000000000000f79d6afbb6da890132f9d7c355e3015f15f3406f
# address 0x0000000000000000000000000000000000000000000000000000000000000000

```

mattsse commented:
what's missing here @zerosnacks ?

louisbrown0212 commented on crates/cast/bin/cmd/creation_code.rs:
```diff
@@ -0,0 +1,159 @@
+    }
+}
+
+/// Parses the creation bytecode to return either the bytecode, or bytecoe without constructor
```
typo: `bytecoe` -> `bytecode`

zerosnacks commented on crates/cast/bin/args.rs:
```diff
@@ -898,6 +898,14 @@ pub enum CastSubcommand {
+    #[command(visible_alias = "cc")]
+    CreationCode(CreationCodeArgs),
+
+    /// Display args used for the contract initialization
+    #[command(visible_alias = "cra")]
+    CreationArgs(CreationArgsArgs),
```
Does this always refer to constructor arguments?

If so, can we rename it `ConstructorArgs`?

zerosnacks commented on crates/cast/bin/cmd/creation_code.rs:
```diff
@@ -0,0 +1,159 @@
+    pub async fn run(self) -> Result<()> {
+        let Self { contract, etherscan, rpc, disassemble, without_args, only_args } = self;
+
+        if without_args && only_args {
+            return Err(eyre!("--without-args and --only-args are mutually exclusive."));
+        }
```
Use the `conflicts_with = "foobarbaz"` argument in `#[arg(long)]` of either to indicate that the two flags are conflicting. Note you only have to do this for one of the args, the other is inferred.

zerosnacks commented on crates/cast/bin/cmd/creation_args.rs:
```diff
@@ -0,0 +1,81 @@
+}
+
+/// Fetches the constructor arguments values and types from the creation bytecode and ABI.
+async fn parse_creation_args(
```
```suggestion
async fn parse_constructor_args(
```

zerosnacks commented on crates/cast/bin/cmd/creation_code.rs:
```diff
@@ -0,0 +1,159 @@
+        })?;
+
+        for trace in traces {
+            if let Some(TraceOutput::Create(CreateOutput { address, code: _, gas_used: _ })) =
```
```suggestion
            if let Some(TraceOutput::Create(CreateOutput { address, .. })) =
```

zerosnacks commented on crates/cast/bin/cmd/creation_code.rs:
```diff
@@ -0,0 +1,159 @@
+            {
+                if address == contract {
+                    creation_bytecode = match trace.trace.action {
+                        Action::Create(CreateAction { init, value: _, from: _, gas: _ }) => {
```
```suggestion
                        Action::Create(CreateAction { init, .. }) => {
```

zerosnacks reviewed (changes requested):
Resolved some nits I found

I think it makes sense to change `creation_args` to `constructor_args` as that is the common way to refer to it

zerosnacks commented on crates/cast/bin/cmd/creation_args.rs:
```diff
@@ -0,0 +1,81 @@
+        .enumerate()
+        .map(|(i, arg)| {
+            let arg = arg.to_vec();
+            format!("{} {}", constructor.inputs[i].ty, Bytes::from(arg))
```
Can we format this as the type passed?

```
            format!("{} {}", constructor.inputs[i].ty, Bytes::from(arg))
```

Should preferably yield

address 0xf79d6afbb6da890132f9d7c355e3015f15f3406f
address 0x00000000000000000000000000000000000000

Instead of 

```
address 0x000000000000000000000000f79d6afbb6da890132f9d7c355e3015f15f3406f
address 0x0000000000000000000000000000000000000000000000000000000000000000
```

pawurb commented:
@zerosnacks applied the review fixes. I've also added `abi_path` argument. It allows to use `constructor-args` for non verified contracts. I've implemented display in this format:

```
0x000000000000000000000000f79d6afbb6da890132f9d7c355e3015f15f3406f -> Address(0xf79d6afbb6da890132f9d7c355e3015f15f3406f)
0x0000000000000000000000000000000000000000000000000000000000000000 -> Address(0x0000000000000000000000000000000000000000)
```
Showing both raw and decoded values can be helpful for Etherscan contract verification that requires raw argument values.

zerosnacks commented on crates/cast/bin/cmd/constructor_args.rs:
```diff
@@ -0,0 +1,104 @@
+    let bytes = Bytes::from(arg.clone());
+    let decoded = arg_type.abi_decode(&arg)?;
+
+    Ok(format!("{bytes} -> {decoded:?}"))
```
```suggestion
    Ok(format!("{bytes} → {decoded:?}"))
```

zerosnacks commented on crates/cast/bin/args.rs:
```diff
@@ -937,6 +938,14 @@ pub enum CastSubcommand {
+    #[command(visible_alias = "cc")]
+    CreationCode(CreationCodeArgs),
+
+    /// Display args used for the contract initialization
```
```suggestion
    /// Display constructor arguments used for the contract initialization.
```

zerosnacks commented on crates/cast/bin/cmd/constructor_args.rs:
```diff
@@ -0,0 +1,104 @@
+}
+
+fn format_arg(ty: &str, arg: Vec<u8>) -> Result<String> {
+    let arg_type: DynSolType = ty.parse().expect("Invalid ABI type.");
```
Elegant solution w/ DynSolType :+1:

zerosnacks commented on crates/cast/bin/cmd/creation_code.rs:
```diff
@@ -0,0 +1,165 @@
+    /// An Ethereum address, for which the bytecode will be fetched.
+    contract: Address,
+
+    /// Path to file containing the contract's JSON ABI. It's necessary if the target contract is
+    /// not verified on Etherscan
```
```suggestion
    /// Path to file containing the contract's JSON ABI. It's necessary if the target contract is
    /// not verified on Etherscan.
```

grandizzy commented on crates/cast/bin/cmd/constructor_args.rs:
```diff
@@ -0,0 +1,104 @@
+
+        let args_arr = parse_constructor_args(bytecode, contract, &etherscan, abi_path).await?;
+        for arg in args_arr {
+            println!("{arg}");
```
I think we should switch to from `println!` to `sh_println!` macro we introduced recently

mattsse commented on crates/cast/bin/cmd/constructor_args.rs:
```diff
@@ -0,0 +1,104 @@
+    let abi = abi.into_iter().next().ok_or_eyre("No ABI found.")?;
+    let (abi, _) = abi;
+
+    if abi.constructor.is_none() {
+        return Err(eyre!("No constructor found."));
+    }
+
+    let constructor = abi.constructor.unwrap();
```
nit: this can be `abi.constructor.ok_or_else?` I think

mattsse commented on crates/cast/bin/cmd/constructor_args.rs:
```diff
@@ -0,0 +1,104 @@
+
+fn format_arg(ty: &str, arg: Vec<u8>) -> Result<String> {
+    let arg_type: DynSolType = ty.parse().expect("Invalid ABI type.");
+    let bytes = Bytes::from(arg.clone());
+    let decoded = arg_type.abi_decode(&arg)?;
```
this alloc is likely not required and the function can also accept &[u8]

mattsse commented on crates/cast/bin/cmd/constructor_args.rs:
```diff
@@ -0,0 +1,104 @@
+        .chunks(32)
+        .enumerate()
+        .map(|(i, arg)| {
+            let arg = arg.to_vec();
```
then we also dont need the to_vec here I believe

mattsse reviewed (approved):
pedantic style nits, otherwise lgtm

grandizzy reviewed (approved):
lgtm, thank you. I left a comment re using newly introduced `sh_println` macro
